### PR TITLE
ABS, DBLE, SIGN, MOD, ICHAR intrinsic lowering

### PR DESCRIPTION
**Pull request by jeanPerier**
Implement lowering a batch of intrinsic used in BLAS.
LEN_TRIM is still outstanding (needs runtime).

### CODE REVIEW

File: flang/lib/Lower/Intrinsics.cpp [hunk 1]
```diff
@@ -625,13 +661,63 @@ IntrinsicLibrary::Implementation::genRuntimeCall(Context &context,
     }
   } else {
     // could not find runtime function
+    llvm::errs() << "missing intrinsic: " << context.name << "\n";
     assert(false && "no runtime found for this intrinsics");
```
Comment (schweitzpgi) [inline flang/lib/Lower/Intrinsics.cpp:665]: You could change this `assert(false &&` to `llvm_unreachable(`.

File: flang/lib/Lower/Intrinsics.cpp [hunk 2]
```diff
@@ -625,13 +661,63 @@ IntrinsicLibrary::Implementation::genRuntimeCall(Context &context,
...
+}
+
+// ABS
+mlir::Value
+IntrinsicLibrary::Implementation::genAbs(Context &genCtxt,
+                                         MathRuntimeLibrary &runtime) {
+  assert(genCtxt.arguments.size() == 1);
+  auto arg = genCtxt.arguments[0];
+  auto type = arg.getType();
+  if (type.isa<mlir::FloatType>() || type.isa<fir::RealType>()) {
```
Comment (schweitzpgi) [inline flang/lib/Lower/Intrinsics.cpp:688]: I think we have some generic type test floating around now...

Comment (jeanPerier) [reply flang/lib/Lower/Intrinsics.cpp:688]: I agree we should, I could not find one (we have one or two local implementations). I created a isa_real and isa_complex (replacing FIROpsBuilder::isComplex) inside FIRType.h.

**Comment by schweitzpgi** [comment]
Looks good.

File: flang/lib/Lower/Intrinsics.cpp (hunk 2 above)
Comment (schweitzpgi) [reply flang/lib/Lower/Intrinsics.cpp]: That's a nice improvement. Moves more FIR type related stuff to the same place.

**Pull request merged**